akka-stream/src/main/scala/akka/stream/io/impl/InputStreamPublisher.scala

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.io.impl
+
+import java.io.InputStream
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+import akka.util.ByteString.ByteString1C
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object InputStreamPublisher {
+
+  def props(is: InputStream, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int): Props = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[InputStreamPublisher], is, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  private final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class InputStreamPublisher(is: InputStream, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  // TODO possibly de-duplicate with SynchronousFilePublisher?
+
+  import InputStreamPublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
+
+  override def preStart() = {
+    try {
+      readAndSignal(initialBuffer)
+    } catch {
+      case ex: Exception ⇒
+        onError(ex)
+    }
+
+    super.preStart()
+  }
+
+  def receive = {
+    case ActorPublisherMessage.Request(elements) ⇒ readAndSignal(maxBuffer)
+    case Continue                                ⇒ readAndSignal(maxBuffer)
+    case ActorPublisherMessage.Cancel            ⇒ context.stop(self)
+  }
+
+  def readAndSignal(readAhead: Int): Unit =
+    if (isActive) {
+      // signal from available buffer right away
+      signalOnNexts()
+
+      // read chunks until readAhead is fulfilled
+      while (availableChunks.length < readAhead && !eofEncountered && isActive)
+        loadChunk()
+
+      if (totalDemand > 0) self ! Continue
+      else if (availableChunks.isEmpty) signalOnNexts()
+    }
+
+  @tailrec private def signalOnNexts(): Unit =
+    if (availableChunks.nonEmpty) {
+      if (totalDemand > 0) {
+        val ready = availableChunks.head
+        availableChunks = availableChunks.tail
+
+        onNext(ready)
+
+        if (totalDemand > 0) signalOnNexts()
+      }
+    } else if (eofEncountered) onComplete()
+
+  /** BLOCKING I/O READ */
+  def loadChunk() = try {
+    val arr = Array.ofDim[Byte](chunkSize)
+
+    // blocking read
+    val readBytes = is.read(arr)
+
+    readBytes match {
+      case -1 | 0 ⇒
```

Comment:
The [Javadoc of InputStream.read](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[],%20int,%20int)) say this:

> This method blocks until input data is available, end of file is detected, or an exception is thrown.

> If len is zero, then no bytes are read and 0 is returned; otherwise, there is an attempt to read at least one byte. If no byte is available because the stream is at end of file, the value -1 is returned; otherwise, at least one byte is read and stored into b.

So, I'd agree with @huntc that reading zero should either be treated as an InvalidStateException or just be ignored to be lenient towards almost working implementations of `InputStream`.